include/perfetto/tracing/internal/track_event_data_source.h

```diff
@@ -220,53 +220,37 @@ struct TrackEventDataSourceTraits : public perfetto::DefaultDataSourceTraits {
   }
 };
 
-// A generic track event data source which is instantiated once per track event
-// category namespace.
-template <typename DerivedDataSource,
-          const TrackEventCategoryRegistry* Registry>
-class TrackEventDataSource
-    : public DataSource<DerivedDataSource, TrackEventDataSourceTraits> {
-  using Base = DataSource<DerivedDataSource, TrackEventDataSourceTraits>;
+// A generic track event data source.
+class PERFETTO_EXPORT_COMPONENT TrackEventDataSource
+    : public DataSource<TrackEventDataSource, TrackEventDataSourceTraits> {
+  using Base = DataSource<TrackEventDataSource, TrackEventDataSourceTraits>;
 
  public:
   static constexpr bool kRequiresCallbacksUnderLock = false;
 
-  // Add or remove a session observer for this track event data source. The
-  // observer will be notified about started and stopped tracing sessions.
-  // Returns |true| if the observer was successfully added (i.e., the maximum
-  // number of observers wasn't exceeded).
-  static bool AddSessionObserver(TrackEventSessionObserver* observer) {
-    return TrackEventInternal::AddSessionObserver(*Registry, observer);
-  }
-
-  static void RemoveSessionObserver(TrackEventSessionObserver* observer) {
-    TrackEventInternal::RemoveSessionObserver(*Registry, observer);
-  }
-
   // DataSource implementation.
   void OnSetup(const DataSourceBase::SetupArgs& args) override {
     auto config_raw = args.config->track_event_config_raw();
     bool ok = config_.ParseFromArray(config_raw.data(), config_raw.size());
     PERFETTO_DCHECK(ok);
-    TrackEventInternal::EnableTracing(*Registry, config_, args);
+    TrackEventInternal::EnableTracing(config_, args);
   }
 
-  void OnStart(const DataSourceBase::StartArgs& args) override {
-    TrackEventInternal::OnStart(*Registry, args);
-  }
+  void OnStart(const DataSourceBase::StartArgs& args) override;
 
   void OnStop(const DataSourceBase::StopArgs& args) override {
     auto outer_stop_closure = args.HandleStopAsynchronously();
     StopArgsImpl inner_stop_args{};
     uint32_t internal_instance_index = args.internal_instance_index;
     inner_stop_args.internal_instance_index = internal_instance_index;
-    inner_stop_args.async_stop_closure = [internal_instance_index,
+    auto registries = TrackEventInternal::GetRegistries();
```

Comment:
Thanks for fixing the race condition at OnSetup, with EnableTracing.

I think we should fix the same race condition here and in other places as well.

Basically everywhere we use `TrackEventInternal::GetRegistries()` (lookin at the current use cases I think the function should be removed).

That means we will call user callbacks of Observer under the mutex, but that should be ok, you shouldn't add new registries in those. It's probably a good idea to use an std::recursive_mutex, for this reason.